File: lib/src/main/java/io/github/kory33/guardedqueries/core/utils/datastructures/ImmutableStack.java
package io.github.kory33.guardedqueries.core.utils.datastructures;

import java.util.Arrays;
import java.util.Collections;
import java.util.Iterator;
import java.util.NoSuchElementException;

public interface ImmutableStack<T> extends Iterable<T> {
    record Cons<T>(T head, ImmutableStack<T> tail) implements ImmutableStack<T> {
        @Override
        public Iterator<T> iterator() {
            return new Iterator<T>() {
                private ImmutableStack<T> current = Cons.this;

                @Override
                public boolean hasNext() {
                    return !current.isEmpty();
                }

                @Override
                public T next() {
                    if (!hasNext()) {
                        throw new NoSuchElementException("Nil");
                    }
                    Cons<T> currentCons = (Cons<T>) current;
                    current = currentCons.tail;
                    return currentCons.head;
                }
            };
        }

        @Override
        public String toString() {
            final var builder = new StringBuilder();
            builder.append("ImmutableStack(");
            for (T item : this) {
                builder.append(item).append(", ");
            }
            builder.delete(builder.length() - 2, builder.length());
            builder.append(")");
            return builder.toString();
        }
    }

    class Nil<T> implements ImmutableStack<T> {
        private Nil() {
        }

        // we will share this singleton instance
        private static final Nil<?> INSTANCE = new Nil<>();

        public static <T> Nil<T> getInstance() {
            //noinspection unchecked
            return (Nil<T>) INSTANCE;
        }

        @Override
        public Iterator<T> iterator() {
            return Collections.emptyIterator();
        }

        @Override
        public String toString() {
            return "ImmutableStack()";
        }
    }

    default ImmutableStack<T> push(T item) {
        return new Cons<>(item, this);
    }

    default boolean isEmpty() {
        // equivalent to "this instanceof Nil" because Nil only has one instance
        return this == empty();
    }

    default T unsafeHead() {
        if (this.isEmpty()) {
            throw new NoSuchElementException("Nil");
        }

        return ((Cons<T>) this).head;
    }

    default ImmutableStack<T> dropHead() {
        if (this.isEmpty()) {
            return empty();
        } else {
            return ((Cons<T>) this).tail;
        }
    }

    default ImmutableStack<T> replaceHeadIfNonEmpty(T item) {
        if (this.isEmpty()) {
            return empty();
        } else {
            return this.dropHead().push(item);
        }
    }

    default ImmutableStack<T> reverse() {
        // fromIterable puts the first element of the iterable at the bottom of the stack
        // and the iterator implementation of ImmutableStack traverses the stack from the top to the bottom
        return fromIterable(this);
    }

    static <T> ImmutableStack<T> fromIterable(Iterable<T> iterable) {
        ImmutableStack<T> current = Nil.getInstance();
        for (T item : iterable) {
            current = current.push(item);
        }
        return current;
    }

    @SafeVarargs
    static <T> ImmutableStack<T> of(T... items) {
        return fromIterable(Arrays.asList(items));
    }

    static <T> ImmutableStack<T> empty() {
        return Nil.getInstance();
    }
}

File: lib/src/main/scala/io/github/kory33/guardedqueries/core/utils/datastructures/ImmutableStack.scala
package io.github.kory33.guardedqueries.core.utils.datastructures

import java.util
import java.util.Collections
import java.util.NoSuchElementException

// TODO: replace with Scala List
object ImmutableStack {
  case class Cons[T](head: T, tail: ImmutableStack[T]) extends ImmutableStack[T] {
    override def iterator: util.Iterator[T] = new util.Iterator[T]() {
      private var current: ImmutableStack[T] = Cons.this

      override def hasNext: Boolean = !current.isEmpty

      override def next: T = {
        if (!hasNext) throw new NoSuchElementException("Nil")
        val oldCons = current.asInstanceOf[Cons[T]]
        current = oldCons.tail
        oldCons.head
      }
    }

    override def toString: String =
      import scala.jdk.CollectionConverters._
      s"ImmutableStack(${this.iterator.asScala.mkString(", ")}))"
  }

  object Nil {
    // we will share this singleton instance
    private val INSTANCE = new ImmutableStack.Nil[AnyRef]
    def getInstance[T]: ImmutableStack.Nil[T] = {
      // noinspection unchecked
      INSTANCE.asInstanceOf[ImmutableStack.Nil[T]]
    }
  }

  class Nil[T] private extends ImmutableStack[T] {
    override def iterator: util.Iterator[T] = Collections.emptyIterator
    override def toString = "ImmutableStack()"
  }

  def fromIterable[T](iterable: java.lang.Iterable[T]): ImmutableStack[T] = {
    var current: ImmutableStack[T] = Nil.getInstance
    import scala.jdk.CollectionConverters._
    for (item <- iterable.asScala) { current = current.push(item) }
    current
  }

  @SafeVarargs def of[T](items: T*): ImmutableStack[T] =
    import scala.jdk.CollectionConverters._
    fromIterable(items.asJava)

  def empty[T]: ImmutableStack[T] = Nil.getInstance
}

trait ImmutableStack[T] extends java.lang.Iterable[T] {
  def push(item: T) = new ImmutableStack.Cons[T](item, this)

  def isEmpty: Boolean =
    // equivalent to "this instanceof Nil" because Nil only has one instance
    this eq ImmutableStack.empty

  def unsafeHead: T =
    if (this.isEmpty) throw new NoSuchElementException("Nil")
    else this.asInstanceOf[ImmutableStack.Cons[T]].head

  def dropHead: ImmutableStack[T] =
    if (this.isEmpty) ImmutableStack.empty
    else this.asInstanceOf[ImmutableStack.Cons[T]].tail

  def replaceHeadIfNonEmpty(item: T): ImmutableStack[T] =
    if (this.isEmpty) ImmutableStack.empty
    else this.dropHead.push(item)

  def reverse: ImmutableStack[T] = {
    // fromIterable puts the first element of the iterable at the bottom of the stack
    // and the iterator implementation of ImmutableStack traverses the stack from the top to the bottom
    ImmutableStack.fromIterable(this)
  }
}
